bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,135 @@
+![Corda](https://www.corda.net/wp-content/uploads/2016/11/fg005_corda_b.png)
+
+# CordApp distribution service (CDS)
+
+DOCUMENT MANAGEMENT
+---
+
+## Document Control
+
+| Title                | CordApp distribution service                             |
+| -------------------- | ------------------------------------------------------------ |
+| Date                 | 22 Aug 2018                                                |
+| Author               | Ivan Schasny, Mike Hearn |
+| Distribution         | Design Review Board, Product Management, Solutions Engineering, Platform Delivery |
+| Corda target version | OS                                                   |
+
+## HIGH LEVEL DESIGN
+
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute CorDapps updates to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios, when upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/message/190?p=,,,20,0,0,0::relevance,,updates+distribution,20,2,0,22686107):
+
+* BNO should be able to notify members about new CorDapp version availability
+* BNO should be able to revoke a specific version of CorDapp from their BN
+* A member should be able to download a specific version of CorDapp
+* BNO should be able to collect a report from members about CorDapp versions installed on their nodes
+* Members should be able to subscribe to a repository channel and be aware of it's synchronicity
+* It should be possible to prevent a CorDapp from working if a newer version is available
+* CDS should support release channels
```

Comment:
what does this mean?

Reply:
This was originally raised by Fuzz and also mentioned quite a lot in the discussion on groups.io. Users wanted to have something akin to Chrome release channels, i.e. to be able to distribute alfa / beta / release / etc. versions of a CorDapp to their networks. I.e. a user should be able to "subscribe" to a "beta" channel and install "beta" versions of a CorDapp to his node if he wishes so.  I expanded this further down in the design 

```
CDS won't natively support *release channels*. Release channels can be simulated by using multiple Maven repositories (a repo per channel). Users can sync down different remote repositories to different local locations and then manually install a desired version of CorDapp to their nodes.
```